### ISSUE TITLE
Prove Bézout's identity in iset.mm

### ISSUE THREAD

**Issue by jkingdon** [metamath/set.mm#2399]
The set.mm proof (which at a glance would appear to be similar to the one at https://en.wikipedia.org/wiki/B%C3%A9zout's_identity#Proof ) relies on being able to take the infimum of `{𝑧 ∈ ℕ ∣ ∃𝑥 ∈ ℤ ∃𝑦 ∈ ℤ 𝑧 = ((𝐴 · 𝑥) + (𝐵 · 𝑦))}`. Our first impulse might be to formalize this using `zsupcl` (already in iset.mm), or more conveniently a new theorem which combines `zsupcl` with something more or less like `infssuzcl` from set.mm, but only if we could show that `∃𝑥 ∈ ℤ ∃𝑦 ∈ ℤ 𝑧 = ((𝐴 · 𝑥) + (𝐵 · 𝑦))` is decidable. Unless there is some shortcut I'm not seeing, to prove this would require much of the work or perhaps more (that is, decidability of this expression means giving a yes/no answer to whether a number has cofactors, whereas Bézout's identity is saying that the answer for the greatest common divisor is "yes").

The more promising approach would appear to be via https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm#Proof or the "Proof of Bézout's Identity" section of https://brilliant.org/wiki/bezouts-identity/ . Both of these do look a bit daunting in terms of how much there is to formalize, but they at least do provide a way to compute the cofactors which apparently could be turned into a proof by induction in the now-familiar style used in zsupcllemex (currently in iset.mm) and others which prove something of the form `E. ? ??`  via induction.

Any other suggestions or references to consult? I have a number theory textbook sitting at home which I haven't tried yet (but won't be back there for roughly a week, and I don't know whether it would help anyway).

**Comment by digama0** [comment]
I think the extended euclidean algorithm is the way to go. The lean proof is https://github.com/leanprover-community/mathlib/blob/f80c18bfdf620b6fcf462430136c61d0ded07035/src/data/int/gcd.lean#L102 which is completely constructive. Rather than defining the `gcd_a` and `gcd_b` functions here, I would just prove that there exists a value satisfying the properties of `xgcd_aux` there.

**Comment by jkingdon** [comment]
Here's my attempt to write up the proof using a combination of metamath notation and informal conventions which are hopefully clear.

Define F to be a function which takes six arguments r0 s0 t0 r1 s1 t1 (where r0,r1 e. NN and s0, t0, s1, t1 e. ZZ) and returns three r,s,t (where r e. NN and s,t e. ZZ). If r0 is 0 then the value of F is `<. r1 , s1 , t1 >.` If r0 is not 0 then define

```
divide r1 by r0 to get quotient q and remainder m
note that m < r0 so the following line is possible by strong recursion
F(r0,s0,t0,r1,s1,t1) := F(m, s1 - q x. s0 , t1 - q x. t0 , r0 , s0 , t0 )
```

Recall that our goal is `((𝐴 ∈ ℤ ∧ 𝐵 ∈ ℤ) → ∃𝑥 ∈ ℤ ∃𝑦 ∈ ℤ (𝐴 gcd 𝐵) = ((𝐴 · 𝑥) + (𝐵 · 𝑦)))`. So given `F` above let <. z , x , y >. be `F(A , 1 , 0 , B , 0 , 1)`. `x` and `y` are the numbers we want and we just need to prove `(𝐴 gcd 𝐵) = ((𝐴 · 𝑥) + (𝐵 · 𝑦)))`.

I'm not sure that quite ends with "Q.E.D." but am I even close? I feel like (a) the above might have "off by one iteration" or other errors, (b) there's a moderate amount of "how do I even express this in iset.mm?" in the above, and probably most importantly, (c) we want to recast this as induction rather than recursion, right? Because that's how we are going to be able to prove something at each step which will eventually turn into `(𝐴 gcd 𝐵) = ((𝐴 · 𝑥) + (𝐵 · 𝑦)))`, right? And it might be easier? (or is that not what @digama0 had in mind by "just prove that there exists a value satisfying the properties of xgcd_aux"?) The invariant at each stage is some flavor of `r = A x. s + B x. t` (see `private def P` in the mathlib proof) and that this remains true when we iterate is expressed by the two lines above "Thus s_{k} and t_{k} are Bézout coefficients" in https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm#Proof and by `simp [mul_add, mul_comm, mul_left_comm, add_comm, add_left_comm, sub_eq_neg_add, mul_assoc]` in mathlib.

**Comment by digama0** [comment]
I think that will work, but I had in mind something slightly different for accomplishing (c). Define `( A gcd B )` as `( iota d A. z ( z || d <-> ( z || A /\ z || B ) ) )` (there are some omitted typing conditions), and we will prove by induction that this iota expression exists (and further satisfies Bezout's identity).

Fix `A` and `B`. Let `is_bezout(r)` mean `E. s E. t r = A * s + B * t` (this will be the analogue of the mathlib `P` function), and let `sub_gcd(d, x, y)` mean `A. z ( z || d -> ( z || x /\ z || y ) )` (and `is_gcd` is the biconditional version). We want to prove by strong induction on `x` generalizing `y` that `is_bezout(x) /\ is_bezout(y) -> E. r ( sub_gcd(r, x, y) /\ is_bezout(r) )`. (Separately, you can prove `E* r is_gcd(r, A, B)`, and `is_bezout(r)` implies the reverse direction of `sub_gcd`, so that the `r` being talked about in this theorem is in fact `( A gcd B )`.) To prove the final theorem we set `x = A` and `y = B` since `is_bezout(A)` and `is_bezout(B)` are trivial.

* In the base case, we have `x = 0`, so take `r = y` and `sub_gcd(y, 0, y)` is trivial.
* In the inductive case, suppose `is_bezout(x) /\ is_bezout(y)`, and let `x' := y % x` and `y' := x`, then `x' < x`, `is_bezout(y')` holds trivially and `is_bezout(x')` holds (1), so the IH applies; and `sub_gcd(r, x', y') -> sub_gcd(r, x, y)` by some more algebra (2).

To elaborate on (1): we know `x = A * s + B * t` and `y = A * s' + B * t'`, so let `y = q * x + x'` and then `x' = A * (s' - q*s) + B * (t' - q*t)`.

To elaborate on (2): If `z || r`, then `z || x'` and `z || x` by assumption, so `z || (q * x + x')` so `z || y`.

**Comment by jkingdon** [comment]
> I had in mind something slightly different

Nice. It is a lot easier to see how that maps to metamath notation. I'll see how it goes once I start formulating and proving the various lemmas, but at least to look over it, it seems like it should be easy to get started once I find some time.

**Comment by jkingdon** [comment]
> We want to prove by strong induction on x generalizing y that is_bezout(x) /\ is_bezout(y) -> E. r ( sub_gcd(r, x, y) /\ is_bezout(r) )

How does this work? I'll start with not being familiar with the phrase "strong induction on x generalizing y". And then I'll ask how we can induct on x - is this proof only for x >= 0? I would have thought that d,r e. NN and s,t,A,B,x,y,z e. ZZ but I could be wrong about some of them.

**Comment by digama0** [comment]
By "strong induction on x generalizing y" I mean that we are proving `A. x e. NN0 ( x < t -> A. y e. NN0 P(x, y) )` by induction on `t e. NN0`. Assuming you have a proof schema for strong induction, i.e. `A. t e. NN0 ( A. x e. NN0 ( x < t -> P(x) ) -> P(t) ) -> A. n e. NN0 P(n)`, this simplifies to just proving `A. y e. NN0 P(x, y)` by strong induction on `x e. NN0`.

The theorem itself is only being proven for nonnegative x, y; for negative values we just use ``( A gcd B ) = ( ( abs ` A ) gcd ( abs ` B ) )``, and Bezout's identity is extended to integers [by cases](https://github.com/leanprover-community/mathlib/blob/268d1a8620537cf9f7304c9f4caceac9e7c1f785/src/data/int/gcd.lean#L142). In the mathlib proof `d,r,x,y,A,B e. NN0` and `s,t e. ZZ`.

**Comment by digama0** [comment]
I think you can more or less trivially take `y e. ZZ` in the proof above; you can take `x e. ZZ` as well but in that case you are doing induction on `abs(x)` instead. You need a sensible definition of `y % x` and `q := fl(y / x)` in the case where `x` and/or `y` are negative; the relevant fact for the recursion to decrease is that `0 <= y % x < abs(x)` as long as `x != 0`.

**Comment by jkingdon** [comment]
> Assuming you have a proof schema for strong induction

We do now: uzsinds (and nnsinds and nn0sinds) was merged as part of #2407 

> The theorem itself is only being proven for nonnegative x, y . . . In the mathlib proof d,r,x,y,A,B e. NN0 and s,t e. ZZ.

Ah, got it. It is `s` and `t` which need to be able to be negative for the whole thing to work. Keeping `x` and `y` as nonnegative for the induction and then proving the theorem for all integers in terms of the one for nonnegative ones, as mathlib does, seems like the most appealing approach.

I've opened a draft pull request at #2413 which I'll keep modifying as I start to prove some of this. Right now it just states a few of the lemmas.

**Comment by jkingdon** [comment]
> Define `( A gcd B )` as `( iota d A. z ( z || d <-> ( z || A /\ z || B ) ) )`

Did you have a proof in mind for this? In set.mm this theorem (stated as http://us.metamath.org/mpeuni/dfgcd2.html which is different in various details but looks similar at least at first glance) is proved in terms of http://us.metamath.org/mpeuni/bezout.html (via http://us.metamath.org/mpeuni/dvdsgcd.html ).

**Comment by digama0** [comment]
@jkingdon What do you mean? That's a definition, it shouldn't need a proof per se (AFAIK `iota` works just fine in iset.mm). If you mean how to prove that `E! d A. z ( z || d <-> ( z || A /\ z || B ) )`, this follows partially from the proof above:

* Bezout's theorem is proving `E. r ( sub_gcd(r, x, y) /\ is_bezout(r) )`
* It follows from `is_bezout(r)` that `sub_gcd(r, x, y) -> is_gcd(r, x, y)`
* Therefore `E. d A. z ( z || d <-> ( z || A /\ z || B ) )`

Separately, we prove `E* d A. z ( z || d <-> ( z || A /\ z || B ) )`: Given `d1, d2` satisfying this, you have `A. z ( z || d1 <-> z || d2 )`, so `d1 || d2` and `d2 || d1`, and divisibility is antisymmetric (this should be a lemma, and it's easy since `x || y -> x <_ y`). (That's only true for nonnegative numbers, but did I mention that `d e. NN0`? The `z` quantifier can either be over `NN0` or `ZZ` by your preference.)

One fun fact about the `iota` definition for gcd is that it does not need a special case for `( 0 gcd 0 )`: `0` is the unique number such that `A. z ( z || d <-> ( z || 0 /\ z || 0 ) )`.

**Comment by jkingdon** [comment]
> That's a definition, it shouldn't need a proof per se

So you are suggesting that we replace the df-gcd which is now in iset.mm (which is the same as the df-gcd in set.mm)? And then prove that this definition proves all the gcd theorems (the ones now in iset.mm and I suppose the ones we plan on bringing over from set.mm)?

I mean, if we have to I guess but the current definition is elegant (it is almost a direct translation into symbols of the words "greatest common divisor") and is what set.mm uses. I'm trying not to be overly influenced by having to throw out all the gcd work I've done so far (since that's just a sunk cost fallacy), but maybe I should ask it this way: what is the purpose of changing the definition? I mean, I don't grasp this well enough to understand the tradeoffs or even what the essential differences (if any) are between the proof I sketch out above, the proof you sketch out above, the proof at https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm#Proof , etc. But maybe say a bit about how the choice of definition affects the ability to prove "the greatest common divisor is then the last non zero remainder r_{{k}}" (in the words of that wikipedia page).

> One fun fact about the iota definition for gcd is that it does not need a special case for ( 0 gcd 0 )

Correct, and this is proved in dfgcd3 which is now in set.mm.

**Comment by digama0** [comment]
@jkingdon One advantage of the definition is that it follows fairly simply from this Bezout proof that the GCD exists (as demonstrated above) and if you recombine this with the Bezout theorem then you get the form you originally wanted: `E. s E. t ( A gcd B ) = ( ( A x. s ) + ( B x. t ) )`. Knowing now that GCD satisfies `C || ( A gcd B ) <-> ( C || A /\ C || B )`, you can then proceed to prove that it is the *greatest* common divisor, in the `<_` order, and hence the sup definition is well defined and the GCD evaluates to that too.

In other words, the proof sketch naturally leads to dfgcd3 being the first definition and df-gcd being the second. It is possible to keep `gcd` itself to use df-gcd, it just means that there will be a few extra lemmas that are like gcd lemmas but talking about the rhs of dfgcd3 first (or alternatively, any `r` satisfying `is_gcd(r, A, B)`, i.e. the thing you get out of bezout), until you have enough to prove that df-gcd was the thing being talked about all along and then you can rephrase the lemmas in terms of it.

**Comment by jkingdon** [comment]
> a few extra lemmas that are like gcd lemmas but talking about the rhs of dfgcd3 first

Ah, I like that. I mean, I might decide to change around (or add) definitions later but for now it seems like working towards

```
    nn0bezout $p |- ( ( A e. NN0 /\ B e. NN0 ) ->
        E. x e. ZZ E. y e. ZZ ( iota_ d e. NN0 A. z e. ZZ
        ( z || d <-> ( z || M /\ z || N ) ) )
        = ( ( A x. x ) + ( B x. y ) ) ) $=
```

should be able to keep things moving. Probably as I get into it I'll see more clearly what you are suggesting in terms of why this iota is going to be important (and what we'll need to prove about it), but I think I'd rather start out with formalizing this statement of `nn0bezout`, and later tackle how the iota relates to the gcd as defined in terms of supremum.

**Comment by digama0** [comment]
My suggestion is to write nn0bezout more like the macro-expanded version of my earlier post:
```
    nn0bezoutlem $p |- ( ( A e. NN0 /\ B e. NN0 ) ->
        E. d e. NN0 ( A. z e. ZZ ( z || d -> ( z || A /\ z || B ) ) /\
          E. x e. ZZ E. y e. ZZ d = ( ( A x. x ) + ( B x. y ) ) ) $=
```
Then there are a few corollaries:
```
    bezoutlem $p |- ( ( A e. ZZ /\ B e. ZZ ) ->
        E. d e. NN0 ( A. z e. ZZ ( z || d <-> ( z || A /\ z || B ) ) /\
          E. x e. ZZ E. y e. ZZ d = ( ( A x. x ) + ( B x. y ) ) ) $=
    gcdex $p |- ( ( A e. ZZ /\ B e. ZZ ) ->
        E. d e. NN0 A. z e. ZZ ( z || d <-> ( z || A /\ z || B ) ) ) $=
```
And then some theorems about the value that satisfies this condition:
```
    gcdlem.1 $e |- ( ph -> A e. ZZ ) $.
    gcdlem.2 $e |- ( ph -> B e. ZZ ) $.
    ${
      gcdlem.3 $e |- ( ph -> D e. NN0 ) $.
      gcdlem.4 $e |- ( ph -> A. z e. ZZ ( z || D <-> ( z || A /\ z || B ) ) ) $.
      ${
        gcdmo.5 $e |- ( ph -> E e. NN0 ) $.
        gcdmo.6 $e |- ( ph -> A. z e. ZZ ( z || E <-> ( z || A /\ z || B ) ) ) $.
        gcdmo $p |- ( ph -> D = E ) $.
      $}
      gcdeu $p |- ( ph -> E! d e. NN0 A. z e. ZZ ( z || d <-> ( z || M /\ z || N ) ) ) $.
      eqgcd3 $p |- ( ph -> D = ( iota_ d e. NN0 A. z e. ZZ ( z || d <-> ( z || M /\ z || N ) ) ) ) $.
      ${
        eqgcd.5 $e |- ( ph -> -. ( A = 0 /\ B = 0 ) ) $.
        eqgcdsuplem $p |- ( ph -> A. z e. ZZ ( ( z || M /\ z || N ) -> z <_ D ) ) $.
        eqgcdsup $p |- ( ph -> D = sup ( { z e. ZZ | ( z || M /\ z || N ) } , < , RR ) ) $.
      $}
      eqgcd $p |- ( ph -> D = ( A gcd B ) ) $.
```
From which dfgcd3, dvdsgcdb and the rest follow.

**Comment by jkingdon** [comment]
> My suggestion is to write nn0bezout more like the macro-expanded version of my earlier post:

Nice. This (and the other lemmas suggested in this comment) formalized nicely (with only some small tweaks), and the result can be seen at #2427 . That includes a proof of dfgcd3 but (as in the above comment), the iota is only used a few places, and other places it is more convenient to work with `E.`.
